draft-ietf-ppm-dap.md

```diff
@@ -2117,29 +2110,43 @@ Helper knows it can clean up its state.
+scenario.
+
 `AggregationJobContinueReq` messages contain a `step` field, allowing
-Aggregators to ensure that their peer is on an expected step of DAP
-aggregation. In particular, the intent is to allow recovery from a scenario
-where the Helper successfully advances from step `n` to `n+1`, but its
-`AggregationJobResp` response to the Leader gets dropped due to something like
-a transient network failure. The Leader could then resend the request to have
-the Helper advance to step `n+1` and the Helper should be able to retransmit
-the `AggregationJobResp` that was previously dropped. To make that kind of
-recovery possible, Aggregator implementations SHOULD checkpoint the most recent
-step's prep state and messages to durable storage such that the Leader can
-re-construct continuation requests and the Helper can re-construct continuation
-responses as needed.
-
-When implementing an aggregation step skew recovery strategy, the Helper SHOULD
-ensure that the Leader's `AggregationJobContinueReq` message did not change when
-it was re-sent (i.e., the two messages must be identical). This prevents the
-Leader from re-winding an aggregation job and re-running an aggregation step
-with different parameters.
-
-One way the Helper could address this would be to store a digest of the Leader's
-request, indexed by aggregation job ID and step, and refuse to service a request
-for a given aggregation step unless it matches the previously seen request (if
-any).
+Aggregators to ensure that their peer is on the expected step of DAP
+aggregation. An aggregation job is said to be "skewed" if the Leader and Helper
+have a different view of the last completed step of aggregation; for example, if
+the Leader does not receive or successfully process an `AggregationJobResp` from
+the Helper, the Leader will believe that the last completed step is one less
+than the Helper. The intent of the `step` field is to allow detection and
+recovery from a skewed aggregation job. When the Leader receives a
+`stepMismatch` error, it SHOULD re-send the original `AggregationJobInitReq`
+message to restart the aggregation process and allow aggregation to complete
+successfully. To make this kind of recovery possible, Leader implementations
+SHOULD retain enough information to reconstruct the `AggregationJobInitReq`
+message in the case that the aggregation process becomes desynchronized.
+
+Changing an aggregation job's parameters is illegal, so the Helper MUST ensure
+that the Leader's `AggregationJobInitReq` message did not change when it was
+re-sent (i.e., the two messages must be identical). This prevents the Leader
+from re-winding an aggregation job and re-running aggregation with different
+parameters.
+
+One way the Helper could implement this would be to store a digest of the
+Leader's request, indexed by aggregation job ID, and refuse to service a re-sent
+request unless it matches the previously seen request.
```

Comment:
FWIW each AggreagtionJobContinueReq needs to be the same as well. Though I believe the protocol logic should ensure this? It's still worth mentioning in any case.

Also the responses shouldn't change either.

Reply:
+1. The Leader could restart an aggregation job and arbitrarily remove valid report shares from subsequent AggregationJobContinueReqs. In that case, the Helper may need to ensure the request digest hasn't changed for every single step involved.

It could do this by storing the digest indexed by aggregation job ID and step. But, this increases the amount of state required from one digest (in the -1 recovery) to n digests (in from-the-beginning recovery).

Reply:
Can y'all spell out the specific problem or attack you have in mind?

I don't think we need to ensure the later requests are the same. My reasoning is:
* If the Helper did not consider the aggregation process complete when the retry was initiated, no state outside of the aggregation job has been modified -- so if the Leader ends up removing a different set of reports on retry, there is no problem.
* If the Helper did consider the aggregation process complete when the retry was initiated, it will have already committed the recovered output shares & noted that it has done so, and thus will not commit a second time. If the Leader aggregates a different set of reports, this will result in a `batchMismatch` error, but the Leader has no ability to e.g. aggregate a slightly different set of reports to infringe on a particular report's privacy.

OTOH, the Leader _accidentally_ aggregating a different set of reports against an aggregation job the Helper considers to be complete is likely a downside of this approach -- it would be hard to avoid in corner cases. The most likely specified corner cases I can find here would be (a) the HPKE key rotating or (b) the replay-checking state being dropped, both of which would result in the report becoming un-aggregatable on retry. (Tracking the exact request/responses at each step wouldn't help -- we'd fail earlier & maybe with a better error code, but would still eventually encounter a `batchMismatch` error on collection, since the Leader would never be able to complete the aggregation job. I think a truly robust implementation of the restart-from-beginning idea would require the Helper to have the ability to un-commit an aggregation job from the running-total aggregate shares when the aggregation job is restarted, which I suppose would require tracking the aggregate shares for each aggregation job -- i.e. we'd lose the storage savings of the running-total aggregation strategy.)

Reply:
Good point. I was originally thinking of something like this: the Helper remembers a hash of each request, and if the request doesn't match the hash, then it aborts the aggregation job. But you're right that we need to be robust to either the Leader or Helper making different rejection decisions.

Another wrinkle pointed out by @divergentdave on the 2024/10/23 on the call: we need to be careful about VDAF preparation. Existing security analysis assumes the attacker can't "rewind" the computation and retry with different inputs. In particular, we're worried that reusing the Prio3 proof shares may result in a privacy violation. (Likewise for the Beaver triples in Poplar1.) In particular we need to ensure the Leader doesn't change the prep share it sends for each report.

Reply:
There was some off-PR discussion about this today, written up in https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/issues/604#issuecomment-2433065034.

The summary is that not only do we need the Helper to verify that the entire sequence of requests is the same on retry, we _also_ need the Leader to verify that the entire sequence of responses is the same on retry.

This is achieved by the restart-from-previous-step currently specified in DAP-12, but is not achieved by the restart-from-beginning (this PR) or restart-from-anywhere (#569) strategies as currently written. We could certainly augment this PR or #569 to specify the requirement that the Leader record the sequence of responses it has successfully processed; but AIUI the original motivation for changing the retry strategy was to permit a Leader implementation which did not need to record anything to durable storage between aggregation steps. If the Leader now has to store response hashes to durable storage between aggregation steps, we no longer meet the original motivation for modifying the retry strategy.

Reply:
Thank you for the detailed write up and I agree. Although this strategy would work for two happy Aggregators, it cannot protect against scenarios where one is malicious, compromising our security guarantees. If the Leader and Helper would need to remember and enforce the exact digests for each step anyways, we may as well just restart from the last "committed" step between the two, which falls into line with the current status quo.